agent/dns.go

```diff
@@ -665,8 +696,37 @@ func (d *DNSServer) doDispatch(network string, remoteAddr net.Addr, req, resp *d
 			goto INVALID
 		}
 
+		lookup := serviceLookup{
+			Network:           network,
+			Datacenter:        datacenter,
+			Service:           queryParts[len(queryParts)-1],
+			Connect:           true,
+			Ingress:           false,
+			MaxRecursionLevel: maxRecursionLevel,
+			EnterpriseMeta:    entMeta,
+		}
 		// name.connect.consul
-		d.serviceLookup(cfg, network, datacenter, queryParts[len(queryParts)-1], "", &entMeta, true, req, resp, maxRecursionLevel)
+		d.serviceLookup(cfg, lookup, req, resp)
+	case "ingress":
+		if len(queryParts) < 1 {
+			goto INVALID
+		}
+
+		if !d.parseDatacenterAndEnterpriseMeta(querySuffixes, cfg, &datacenter, &entMeta) {
+			goto INVALID
+		}
+
+		lookup := serviceLookup{
+			Network:           network,
+			Datacenter:        datacenter,
+			Service:           queryParts[len(queryParts)-1],
+			Connect:           false,
+			Ingress:           true,
```

Comment:
Feels like if you pick up a third boolean you might want to switch to a proper `iota` enum instead of mutually exclusive booleans.